std/js/lib/BufferSource.hx

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C)2005-2019 Haxe Foundation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+package js.lib;
+
+/**
+	`BufferSource` is a typedef used to represent objects that are either themselves an [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer),
+	or which are a [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray) providing an [ArrayBufferView](https://developer.mozilla.org/en-US/docs/Web/API/ArrayBufferView).
+
+	This is a helper type to simplify the specification. It isn't an interface and there are no objects implementing it.
+
+	Documentation [BufferSource](https://developer.mozilla.org/en-US/docs/Web/API/BufferSource) by [Mozilla Contributors](https://developer.mozilla.org/en-US/docs/Web/API/BufferSource$history), licensed under [CC-BY-SA 2.5](https://creativecommons.org/licenses/by-sa/2.5/).
+
+	@see <https://developer.mozilla.org/en-US/docs/Web/API/BufferSource>
+ */
+abstract BufferSource(Dynamic)
```

Comment:
Since BufferSource is a helper to tell us we have an ArrayBuffer, it's useful to have a way to access the ArrayBuffer and fields, for example
```haxe
@:forward
abstract BufferSource(ArrayBuffer)
        to ArrayBuffer from ArrayBuffer {
	private inline function new(arrayBuffer: ArrayBuffer) {
		this = arrayBuffer;
	}

        @:from static public inline function fromView(view: ArrayBufferView) {
		return new BufferSource(view.buffer);
	}
}
```

The advantage is you can now use it like this
```haxe
function upload(data: BufferSource) {
    trace('data length is ${data.byteLength} bytes');
    methodThatRequiresAnArrayBuffer(data);
}
```

Reply:
> Since BufferSource is a helper to tell us we have an ArrayBuffer, it's useful to have a way to access the ArrayBuffer and fields, for example

I added the `buffer` property to BufferSource. Please review it.

> Your current version maps to the spec closely and is correct, however, it's not that useful on it's own in haxe

`BufferSource` is often used in Web API IDL. So I think `BufferSource` should be defined in the standard library. (And I want to fix WebAssebmly extern...)

* https://webassembly.github.io/spec/js-api/#webassemblyinstantiate
* https://www.w3.org/TR/WebCryptoAPI/#subtlecrypto-interface

Reply:
Totally agree BufferSource should be in the std, I [use it myself in typedarrays for the C++ target](https://github.com/haxiomic/gluon/blob/master/typedarray/BufferSource.hx)

You approach makes sense, however it would be nice to avoid the runtime Std.is check when getting the buffer field

The earlier example can be simplified a bit to
```haxe
@:forward
abstract BufferSource(ArrayBuffer)
        to ArrayBuffer from ArrayBuffer {

	@:from static public inline function fromView(view: ArrayBufferView) {
		return cast view.buffer;
	}

}
```

@nadako, what do you think?

Reply:
I'm always all for compile-time checks and this makes sense to me! 👍

Reply:
Thanks your review, nadako and haxiomic!
I changed it.